File: rank_3/level_2/rip.c
#include <string.h>
#include <stdlib.h>
#include <stdio.h>

/* get balance and isvalid are almost the same. start there */

int	get_balance(char *input)
{
	int balance = 0;
	int i = 0;

	while (i < strlen(input))
	{
		if (input[i] == '(')
			balance--;
		if (input[i] == ')')
			balance++;
		i++;
	}
	return (balance);
}

int is_valid(char *input)
{
	int balance = 0;
	int i = 0;

	while (i < strlen(input))
	{
		if (input[i] == '(')
			balance--;
		if (input[i] == ')')
			balance++;
		if (balance > 0)
			return (0);
		i++;
	}
	return (1);
}

void rip(char *input, char *output, int pos, int balance)
{
	if (pos == strlen(input))
	{
		if (balance == 0 && is_valid(output))
			printf("%s\n", output);
		return ;
	}
	if (input[pos] == '(' && balance < 0)
	{
		output[pos] = ' ';
		rip(input, output, pos + 1, balance + 1);
	}
	else if (input[pos] == ')' && balance > 0)
	{
		output[pos] == ' ';
		rip(input, output, pos + 1, balance - 1);
	}
	output[pos] = input[pos];
	rip(input, output, pos + 1, balance);
}

int	main(int argc, char *argv[])
{
	if (argc != 2)
		return (1);
	
	char *output;
	output = malloc(sizeof(char) * strlen(argv[1]));
	rip(argv[1], output, 0, get_balance(argv[1]));
	free(output);
}

File: rank_3/level_2/rip.py
def rip(input: str, output: str, balance: int, pos: int):
	if pos == len(input):
		if balance == 0 and is_valid(output):
			print("".join(output))
		return

	if input[pos] == "(" and balance < 0: #in case we are removing (s
		output[pos] = "_" #remove current
		rip(input, output, balance + 1, pos + 1) #and continue

	if input[pos] == ")" and balance > 0:
		output[pos] = "_"
		rip(input, output, balance - 1, pos + 1)

	output[pos] = input[pos]
	rip(input, output, balance, pos + 1)

def get_balance(input: str):
	balance = 0
	i = 0
	while i < len(input):
		if input[i] == "(":
			balance -= 1
		elif input[i] == ")":
			balance += 1
		i += 1
	return balance

def is_valid(input: str):
	balance = 0
	i = 0
	while i < len(input):
		if input[i] == "(":
			balance -= 1
		elif input[i] == ")":
			balance += 1
		i += 1
		if balance > 0:
			return False
	return True

input = "(()(()("
output = [" "] * len(input)
balance = get_balance(input)
print(f"balance is {balance}")
rip(input, output, balance, 0)

# 5 01234
#   0 23
#     23
#   01  4
